torch/lib/THCUNN/generic/MSECriterion.cu

```diff
@@ -7,12 +7,41 @@ void THNN_(MSECriterion_updateOutput)(
            THCTensor *input,
            THCTensor *target,
            THCTensor *output,
-           bool sizeAverage)
+           bool sizeAverage,
+           bool reduce)
 {
   THCUNN_check_nElement(state, input, target);
-  THCUNN_check_dim_size(state, output, 1, 0, 1);
-  THCUNN_assertSameGPU(state, 2, input, target);
 
+  if (reduce) {
+    THCUNN_check_dim_size(state, output, 1, 0, 1);
+
+    ptrdiff_t size = THCTensor_(nElement)(state, input);
+
+    input = THCTensor_(newContiguous)(state, input);
+    target = THCTensor_(newContiguous)(state, target);
+
+    THCThrustAllocator thrustAlloc(state);
+    thrust::device_ptr<real> input_data(THCTensor_(data)(state, input));
+    thrust::device_ptr<real> target_data(THCTensor_(data)(state, target));
+    accreal sum = thrust::inner_product(
```

Comment:
Wouldn't it be better to use [`THC_pointwiseApply3` instead](https://github.com/pytorch/pytorch/blob/master/torch/lib/THC/generic/THCTensorMathPointwise.cu#L290)? This way we wouldn't need to convert the tensors to contiguous.

Reply:
Oh, just saw that this is for the `reduce=True` case. Disregard my comment. 

I know that this was like this before, but looking at it again I wonder in this case if using [`THC_reduceAll`](https://github.com/pytorch/pytorch/blob/master/torch/lib/THC/generic/THCTensorMathReduce.cu#L142) wouldn't be better.

Reply:
I think I'll leave this as-is because the code used to be like this?

Reply:
Yeah, leave as-is. When I first commented on it, I haven't seen that this was what they had already.